slm_lab/agent/memory/prioritized.py

```diff
@@ -0,0 +1,175 @@
+            j = i + self.capacity - 1
+            print(f'Idx: {i}, Data idx: {self.indices[i]}, Prio: {self.tree[j]}')
+
+
+class PrioritizedReplay(Replay):
+    '''
+    Prioritized Experience Replay
+
+    Implementation follows the approach in the paper "Prioritized Experience Replay", Schaul et al 2015" https://arxiv.org/pdf/1511.05952.pdf and is Jaromír Janisch's with minor adaptations.
+    See memory_util.py for the license and link to Jaromír's excellent blog
+
+    Stores agent experiences and samples from them for agent training according to each experience's priority
+
+    The memory has the same behaviour and storage structure as Replay memory with the addition of a SumTree to store and sample the priorities.
+
+    e.g. memory_spec
+    "memory": {
+        "name": "Replay",
+        "alpha": 1,
+        "epsilon": 0,
+        "batch_size": 32,
+        "max_size": 10000
+    }
+    '''
+
+    def __init__(self, memory_spec, algorithm, body):
+        util.set_attr(self, memory_spec, [
+            'alpha',
+            'epsilon',
+            'batch_size',
+            'max_size',
+        ])
+        self.epsilon = torch.full((1,), self.epsilon)
+        self.alpha = torch.full((1,), self.alpha)
+        super(PrioritizedReplay, self).__init__(memory_spec, algorithm, body)
+
+    def reset(self):
+        super(PrioritizedReplay, self).reset()
+        self.tree = SumTree(self.max_size)
+
+    def add_experience(self, state, action, reward, next_state, done, error=100000):
+        '''
+        Implementation for update() to add experience to memory, expanding the memory size if necessary.
+        All experiences are added with a high priority to increase the likelihood that they are sampled at least once.
+        '''
+        error = torch.zeros(1).fill_(error)
+        priority = self.get_priority(error)
+        super(PrioritizedReplay, self).add_experience(state, action, reward, next_state, done, priority)
+        self.tree.add(priority, self.head)
+
+    def get_priority(self, error):
+        '''Takes in the error of one or more examples and returns the proportional priority'''
+        p = torch.pow(error + self.epsilon, self.alpha)
+        return p.squeeze_().numpy()
+
+    def sample_idxs(self, batch_size):
+        '''Samples batch_size indices from memory in proportional to their priority.'''
+        batch_idxs = np.zeros(batch_size)
+        tree_idxs = np.zeros(batch_size, dtype=np.int)
+        segment = self.tree.total() / batch_size
+
+        for i in range(batch_size):
+            a = segment * i
+            b = segment * (i + 1)
+
+            s = random.uniform(a, b)
+            (tree_idx, p, idx) = self.tree.get(s)
+            batch_idxs[i] = idx
+            tree_idxs[i] = tree_idx
+
+        batch_idxs = np.asarray(batch_idxs).astype(int)
+        self.tree_idxs = tree_idxs
+        # NOTE research: sample latest for CPER
+        # batch_idxs[-1] = self.true_size - 1   # add the latest sample
```

Comment:
See comment below, we could have something like

```
if CER:
    batch_idxs[-1] = self.true_size - 1
```

Makes it easy to do every permutation of DQN with PER and / or CER